File: src/content/docs/topics/cs/leetcode-150/trees/100-same-tree.py
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def build_tree(vals):
    if not vals:
        return None
    root = TreeNode(vals[0])
    q = [root]
    i = 1
    while q and i < len(vals):
        node = q.pop(0)
        if i < len(vals) and vals[i] is not None:
            node.left = TreeNode(vals[i])
            q.append(node.left)
        i += 1
        if i < len(vals) and vals[i] is not None:
            node.right = TreeNode(vals[i])
            q.append(node.right)
        i += 1
    return root

def is_same_tree(p, q):
    pass  # TODO: implement

def _run_tests():
    assert is_same_tree(build_tree([1, 2, 3]), build_tree([1, 2, 3])) == True
    assert is_same_tree(build_tree([1, 2]), build_tree([1, None, 2])) == False
    assert is_same_tree(None, None) == True
    assert is_same_tree(build_tree([1]), None) == False
    assert is_same_tree(build_tree([1]), build_tree([1])) == True
    assert is_same_tree(build_tree([1, 2, 3]), build_tree([1, 2, 4])) == False
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/trees/100-same-tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func isSameTree(p *TreeNode, q *TreeNode) bool {
	// TODO: implement
	return false
}

func runTests() {
	// identical trees
	p1 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
	q1 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
	assert(isSameTree(p1, q1) == true)

	// different structure
	p2 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}}
	q2 := &TreeNode{Val: 1, Right: &TreeNode{Val: 2}}
	assert(isSameTree(p2, q2) == false)

	// both nil
	assert(isSameTree(nil, nil) == true)

	// one nil
	p3 := &TreeNode{Val: 1}
	assert(isSameTree(p3, nil) == false)

	// single node same
	p4 := &TreeNode{Val: 1}
	q4 := &TreeNode{Val: 1}
	assert(isSameTree(p4, q4) == true)

	// different values
	p5 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
	q5 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 4}}
	assert(isSameTree(p5, q5) == false)

	fmt.Println("all tests pass")
}

func main() { runTests() }
